### DISCUSSION TITLE
ESP32 and W5500

### DISCUSSION THREAD

**Discussion by mcsmark** [lathoub/Arduino-AppleMIDI-Library#135] [Q&A] [answered]
I wanted to try the ESP32 with the W5500. In the information you can read : ESP32 with W5500 Setup
And the setup is a link. But the link does not lead to the right page.
So my first question :  is there still some info on how to use this ?
I did install ESP32 and that works.
I also installed AppleMIDI. I read somewhere it would install MIDI lib too but i had to manual install it.
Ok, the example compiled. (it does take very long) And i found some info on how to connect the W5500 module. 
But i do not get an IP address.  So then it halts.
So i am looking for more info :-)

**Comment by lathoub** [comment, marked as answer]
Thanks for telling me the link broke - somehow the original article is no longer available.

Here is what I connected:

```
ESP32      W5500
EN ------> RESET
23 ------> MOSI
19 ------> MISO
18 ------> SCK
 5 ------> CS
```

![image](https://user-images.githubusercontent.com/4082369/137615988-fba37825-1b43-4cc5-bd9c-3a326eaae0ce.png)

In your setup:
```
#include <Ethernet.h>
#define CS 5
...
  Ethernet.init(CS);
  if (Ethernet.begin(mac) == 0) {
    return false;
  }
```

**Reply by mcsmark** [reply]
thank you for your reply.
In the example code there is this :
#define RESET_PIN  26
#define CS_PIN     5
I wonder of this RESET_PIN is the enabled/reset ?  the image suggest that you connected to the RESET/EN pin of the ESP32. And the code seems to use GPIO 26 ?
For me as a beginner it is confusing.

**Reply by lathoub** [reply]
RESET_PIN is used when using the `Ethernet3.h` library, not the OOB `Ethernet.h` (Ethernet3 exposes the reset pin, Ethernet does not)

What is your setup? As a beginner, I recommend the [ESP32 Devkit](https://www.otronic.nl/a-59613972/esp32/esp32-wroom-4mb-devkit-v1-board-met-wifi-bluetooth-en-dual-core-processor/) and this [W5500 board](https://www.bol.com/be/nl/p/otronic-spi-ethernet-module-w5500-arduino/9300000021338127/) and wire them up on a breadboard as indicated.

Good luck!

**Reply by mcsmark** [reply]
thanks for the links. i use similar hardware. i swapped the module and now it works. thanks again for the help.
I connected the RSTN of the W5500 to the GPIO26. 
I like the small latency.  But i think i can have only 2 sessions. Enough to test things. 
I have no idea which ethernet chip could have many sockets/sessions and still have low latency but i understand that this is the wrong place to ask. anyway, thanks again for the connection info. To users that also have problems because they found an image on the wiznet site : do not use it. it does not work. This example works.

**Reply by lathoub** [reply]
> I like the small latency.

Yes, the ESP32 and W5500 result in quite low latency (<1ms) (whilst with a ATMEGA its around 3ms). See [here](https://github.com/lathoub/Arduino-AppleMIDI-Library/wiki/Keeping-Latency-under-control)

> But i think i can have only 2 sessions.

Correct, the amount of sockets on the W5500 is restricted - see discussion [here](https://github.com/lathoub/Arduino-AppleMIDI-Library/wiki/Enlarge-Ethernet-buffer-size-to-avoid-dropping-UDP-packages)

> I have no idea which ethernet chip could have many sockets/sessions and still have low latency

The [W6100 ](https://www.wiznet.io/product-item/w6100/)is better than the W5500 (ipv6 and 8 sockets with 32kb RAM buffer) and works wihout to much hassle in the Arduino ecosphere. (albeit the W6100 is hard to get today)

**Reply by mcsmark** [reply]
thanks again for the help and info. You did a wonderful job with this library and all the examples. good to see that it is supported so well.
i will have a look at w6100 and other wired options.

**Reply by lathoub** [reply]
Keep me posted on your work with the W6100, I want it as well (ipv6!). Thanks @mcsmark

**Reply by lathoub** [reply]
> thank you for your reply.
> In the example code there is this :
> #define RESET_PIN 26
> #define CS_PIN 5
> I wonder of this RESET_PIN is the enabled/reset ? the image suggest that you connected to the RESET/EN pin of the ESP32. And the code seems to use GPIO 26 ?
> For me as a beginner it is confusing.

Indeed better to have an explicit RESET_PIN then connection it to the EN pin from the ESP32

**Comment by lathoub** [comment]
I've noticed that connecting EN (on ESP32) to RESET (on W5500) sometimes caused the W5500 not to boot correctly. Having RESET triggered by the separate pin on the ESP32 is more stable.
